---
title: "Introduction to methlyation and EWAS workshop"
author: "created by Miles Benton"
date: 'Updated: `r format(Sys.time(), "%d %B, %Y")`'
output: 
  html_document:
    df_print: kable
    code_folding: show
    includes:
      in_header: extLogo.html
      after_body: extFooter.html
---

```{r setup, include=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library('printr')
```

-----------  

This is a basic EWAS tutorial, introducing some of the concepts and tools for analysing methylation data.  

The workshop uses public Illumina 450K data from several studies (GSE40279, GSE48472, GSE41114, GSE42700, GSE46573, GSE48472, GSE50586). It is currently designed to investigate the methylation differences between blood (whole blood) and buccal (cheek swabs) tissues. It is aimed at giving a very basic overview of the structure and type of data used in EWAS as well as a few of the available tools and methods to explore this data.  

**Note:** Before starting this workshop please visit [GitHub](https://github.com/sirselim/methylation_EWAS_workshop) and make sure you have downloaded the repository and have set up all the required software for this workshop.  

-----------

## 1. load required packages

We need the `minfi` package installed for this workshop.

If you need to install `minfi` follow the steps below:

```{r, eval=FALSE, message=FALSE, warning=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("minfi")
```

Now load the package:

```{r, message=FALSE, warning=FALSE}
require(minfi)
```  
  
-----------
  
## 2. load data

The data for this workshop has been provided for you. There are 2 main data files:

  1. annotation for the Illumina 450k methylation array
  2. methylation data (beta values) for 5 whole blood and 5 buccal samples

Load this into R:

```{r, message=FALSE, warning=FALSE}
meth.anno <- readRDS('data/Ilmn450K_anno.RDS')
betas <- readRDS('data/blood_buccal.RDS')
```
  
-----------
  
## 3. set up samples/data and basic QC explore

We'll create an object called `tissues` which provides identification of our blood and buccal samples:

```{r, message=FALSE, warning=FALSE}
# define samples
tissues <- c(rep(0, 5), rep(1, 5))
tissues <- as.factor(tissues)
levels(tissues) <- c('Blood', 'Buccal')
```

Now we'll explore the distribution of methylation values (beta values) by creating a density plot. A methylation distribution for a given individual follows a bimodal distribution, with peaks around 0 (unmethylated) and 1 (methylated).

Create a density plot for all samples, what do you see?

```{r, message=FALSE, warning=FALSE, fig.align='center'}
# create a density plot
densityPlot(betas, sampGroups = tissues)
```

Another way of exploring a data set is by cluster analysis.

Here we perform hierarchical clustering on the 10 samples using all available data:

```{r, message=FALSE, warning=FALSE, fig.align='center'}
# Hierarchical Clustering
d <- t(betas)
rownames(d) <- tissues
d <- dist(d, method = "minkowski", p = 2)
hc <- hclust(d)
plot(hc, main="Cluster on all probes", labels = tissues, xlab = '', sub="")
```

What do you conclude from the above?
  
-----------
  
## 4. run dmpFinder - logistic regression

To run our analysis exploring the difference in methylation between blood and buccal we'll use the `dmpFinder` function from the `minfi` package.

Depending on the 'outcome' (trait/phenotype being investigated) `dmpFinder` can be configured for either categorical (logistic regression) or continuous (linear regression). Here we're looking at blood tissue vs buccal tissue, so a categorical trait - we'll be using logistic regression modelling.

Run `dmpFinder`:

```{r, message=FALSE, warning=FALSE}
# minfi dmpfinder
results <- dmpFinder(betas, pheno = tissues, type = "categorical", qCutoff = 1)
results <- na.omit(results)
```

Above we create a `results` object which contains all CpG sites (methylation sites) ranked by p-value (significance), with the most significant sites first. 

Look at the head of `results` (most significant):

```{r, message=FALSE, warning=FALSE}
head(results)
```

We can also explore the tail (least significant):

```{r, message=FALSE, warning=FALSE}
tail(results)
```

The Q-value is an adjustment for multiple correction bias. We're going to extract all CpG sites which have a Q-value less than 0.00005.

First, how many sites will this be?

```{r, message=FALSE, warning=FALSE}
# explore some results
table(results$qval < 0.00005)
```

Now extract these sites into an object called `top.results`:

```{r, message=FALSE, warning=FALSE}
top.results <- results[results$qval <= 0.00005,]
```

For fun we can explore how many sites pass a stringent Bonferroni correction:

**Note:** Bonferroni threshold is `0.05/[all observations]`, so we'll use `0.05/450000` which is `1.1e-7`.

```{r, message=FALSE, warning=FALSE}
# bonferroni correction
table(results$pval <= (0.05/450000))
```
  
-----------
  
## 5. plotting the results

It is important to be able to visualise your data/results, so here we're going to use some of the graphing functions of `R` to do just that.

First lets try the `plotCpg` function from the `minfi` package:

```{r, message=FALSE, warning=FALSE, tidy=TRUE, fig.align='center', fig.show='hold'}
## very basic plotting 
# top CpG marker ('best'/most significantly associated)
plotCpg(betas, cpg = "cg05350268", pheno = tissues, type = "categorical", measure = "beta", ylab = "methylation (beta)", ylim = c(0,1))
# bottom CpG marker ('best'/most significantly associated)
plotCpg(betas, cpg = "cg07309136", pheno = tissues, type = "categorical", measure = "beta", ylab = "methylation (beta)", ylim = c(0,1))
```


We can also use base `R` plotting to create boxplots:

```{r, message=FALSE, warning=FALSE, tidy=TRUE, fig.align='center', fig.show='hold'}
## boxplots
# lets look at a few boxplots
boxplot(betas['cg05350268',] ~ tissues, col = c('darkred', 'cadetblue'), ylim = c(0,1), main = 'cg05350268')
boxplot(betas['cg19653345',] ~ tissues, col = c('darkred', 'cadetblue'), ylim = c(0,1) , main ='cg19653345')
```

If we want to create more than one plot at a time we can do so:

```{r, message=FALSE, warning=FALSE, tidy=TRUE, fig.align='center'}
## multiple boxplots
# loop through a few sites for the top markers (most significantly associated)
tissue.sites <- rownames(top.results)[c(1:9)]

par(mfrow = c(3,3))
for (cpg in tissue.sites) {
  
  boxplot(betas[cpg,] ~ tissues, col = c('darkred', 'cadetblue'), ylab = "methylation (beta)", xlab = "tissue", main = cpg, ylim = c(0,1))
  
}
```

We can also create a pdf containing these plots:

```{r, message=FALSE, warning=FALSE, tidy=TRUE, results = 'hide'}
## multiple boxplots
# can also write this out to a file (pdf)
pdf('figures/blood_v_buccal_topsites_plot.pdf', width = 7, height = 10)
tissue.sites <- rownames(top.results)[c(1:12)]

par(mfrow = c(4,3))
for (cpg in tissue.sites) {
  
  boxplot(betas[cpg,] ~ tissues, col = c('darkred', 'cadetblue'), ylab = "methylation (beta)", xlab = "tissue", main = cpg)
  
}
dev.off()
```

If you navigate to the `figures/` directory you'll find the pdf created above.

### Hierarchical Clustering 2

We previously performed clustering on all CpG sites, now we're going to cluster on those that pass our defined threshold.

First extract a smaller matrix of all significant CpG sites:

```{r, message=FALSE, warning=FALSE, tidy=TRUE}
# subset the beta matrix by the top markers and cluster
sig.betas <- betas[rownames(betas) %in% rownames(top.results),]
```

Now perform the clustering:

```{r, message=FALSE, warning=FALSE, tidy=TRUE, fig.align='center'}
# Hierarchical Clustering 
d <- t(sig.betas)
rownames(d) <- tissues
d <- dist(d, method = "minkowski", p = 2)
hc <- hclust(d)
plot(hc, main="Cluster on selected top tissue discrimination CpG sites", labels = tissues, xlab = '', sub="")
```

What do you see?
  
-----------
  
## 6. prepare and run pathways analysis

Before we can perform a pathways analysis we need to create a list of genes which underlie the CpG sites extracted from our analysis.

First we need to understand what the array annotation looks like:

```{r, message=FALSE, warning=FALSE, tidy=TRUE}
# explore annotation
head(meth.anno) # see what the annotation looks like
```

Now we get a subset of this annotation - this is sub-setting the annotation based on the sites that passed our analysis:

```{r, message=FALSE, warning=FALSE, tidy=TRUE}
# subset annotation based on the top.results
anno.sub <- meth.anno[meth.anno$IlmnID %in% rownames(top.results),]
```

Lets explore this a little. How many 'significant' CpG sites are there per chromosome?

```{r, message=FALSE, warning=FALSE, tidy=TRUE}
# how many different features can be explored?
table(anno.sub$CHR) # number of significant sites per chromosome
```

How many unique genes are there?

```{r, message=FALSE, warning=FALSE, tidy=TRUE}
length(unique(unlist(strsplit(anno.sub$UCSC_RefGene_Name, split = ';')))) # number of unique genes
```

We'll extract these to an object called `genes`:

```{r, message=FALSE, warning=FALSE, tidy=TRUE}
# get the gene list from the provided 450K annotation object
genes <- unique(unlist(strsplit(anno.sub$UCSC_RefGene_Name, split = ';')))
```

Then we can write this to a text file:

```{r, message=FALSE, warning=FALSE, tidy=TRUE}
# write list of genes out to a file
write.table(genes, 'EWAS_blood_buccal_genelist.txt', row.names = F, col.names = F, quote = F)
```

This is the final list of genes which can now be uploaded to your pathways analysis tool.


## Pathways analysis

For this section we will be using a popular web-server tool, WEB-based GEne SeT AnaLysis Toolkit (http://webgestalt.org).

Here is an overview of parameters to set:

  - select `hsapiens` as organism;
  - select `Overrepresentation Enrichment Analysis (ORA)` as method;
  - select `geneontology` AND `Biological_Process` 
    + *NOTE:* after you've run through `Biological_Process` try again with `Cellular_Component` and `Molecular_Function` 
  - for Gene ID select `genesymbol`
  - then use the browse button to select and upload the list of genes you created earlier 
  - for reference set select `genome`
  - in the Advanced parameters section set minimum number of genes to `2` and leave the rest of these settings as default

You're now ready to click the `submit` button and explore your pathways analysis.

Explore the resultant output.

Have a look at the genes that make up some of the enriched pathways, for example does it make sense to see these pathways? Explore some of the functionality of genes such as *CORO1C*, *EGFR*, *LAT*, and *MAEA*, does this support your hypothesis about the role these pathways might be playing?

Have a 'play' with selecting other parameters and seeing what else you can potentially discover about the genes involved in differentiating these two tissue types at a methlyation level.
  
-----------
  
## Extra: Explore sex effects on the data 

This section can be run if there is spare time at the end of the workshop.

### Install/Require packages

```{r, message=FALSE, warning=FALSE}
# install.packages('FactoMineR')
require(FactoMineR)   # used for PCA and plotting
```

### Define and extract sex markers (chrX and Y)

First define how many X and Y chr markers are present on the 450k array:

```{r, message=FALSE, warning=FALSE}
table(meth.anno$CHR == 'chrX' | meth.anno$CHR == 'chrY')
```

These can be extracted:

```{r, message=FALSE, warning=FALSE}
sex.markers <- meth.anno[meth.anno$CHR == 'chrX' | meth.anno$CHR == 'chrY',]$IlmnID
```

Then you can figure out how many of these are in the data that you have loaded:

```{r, message=FALSE, warning=FALSE}
table(rownames(betas) %in% sex.markers)
```

```{r, message=FALSE, warning=FALSE}
sex.betas <- betas[rownames(betas) %in% sex.markers,]
```

Look at the data, can you see any patterns?

```{r, message=FALSE, warning=FALSE}
head(sex.betas)
```

How does it look when clustered?  

```{r, message=FALSE, warning=FALSE, fig.align='center'}
# Hierarchical Clustering 
d <- t(sex.betas)
rownames(d) <- tissues
d <- dist(d, method = "minkowski", p = 2)
hc <- hclust(d)
plot(hc, main="Cluster on X and Y Chr CpG sites", labels = tissues, xlab = '', sub="")
```

This data is obviously mixed sex. We actually know that there are 4 males in this data set so we can create a sex variable to use in further exploratory analysis:

```{r, message=FALSE, warning=FALSE}
sex <- as.factor(c('female', 'female', 'male', 'male', 'female', 'female', 'female', 'male', 'female', 'male'))
```

We are know going to perform a principle components analysis (PCA) as a means to further explore this data:

```{r, message=FALSE, warning=FALSE}
# set up the data for PCA
pca.data <- t(sex.betas)
pca.data <- as.data.frame(pca.data)
pca.data$tissue <- as.factor(tissues)
pca.data$sex <- sex
# perform the PCA
res.pca <- PCA(pca.data, quali.sup = c(ncol(pca.data), ncol(pca.data)-1), graph = FALSE)
```

Create a PCA plot:  
<!-- <input type=button class=hideshow></input> -->
```{r pca_plot, message=FALSE, warning=FALSE, tidy=TRUE, fig.align='center'}
plotellipses(res.pca, keepvar = c('tissue', 'sex'), axes = c(1,2), cex = 0.65, legend = list(bty = "y", x = "topright"))
```

What can you infer from the PCA plots? Does it seem like sex could be influencing the analysis?

We can actually run a PCA on ALL CpG sites, this will help us to decide whether sex is potentially impacting our analysis.

**WARNING:** the below will take a few minutes (~5) to complete.

```{r, eval=FALSE, message=FALSE, warning=FALSE, tidy=TRUE}
# set up the data for PCA
pca.data <- t(betas)
pca.data <- as.data.frame(pca.data)
pca.data$tissue <- as.factor(tissues)
pca.data$sex <- sex
# perform the PCA
res.pca <- PCA(pca.data, quali.sup = c(ncol(pca.data), ncol(pca.data)-1), graph = FALSE)
```

```{r, eval=FALSE, message=FALSE, warning=FALSE, tidy=TRUE, fig.align='center'}
plotellipses(res.pca, keepvar = c('tissue', 'sex'), axes = c(1,2), cex = 0.65, legend = list(bty = "y", x = "topright"))
```

What can we infer about sex and it's influence on our analysis from the above?

<!-- <script> -->
<!-- $( "input.hideshow" ).each( function ( index, button ) { -->
<!--   button.value = 'Toogle figure'; -->
<!--   $( button ).click( function () { -->
<!--     var target = this.nextSibling ? this : this.parentNode; -->
<!--     target = target.nextSibling.nextSibling.nextSibling.nextSibling.nextSibling; -->
<!--     if ( target.style.display == '' || target.style.display == 'block' ) { -->
<!--       target.style.display = 'none'; -->
<!--       this.value = 'Toogle figure'; -->
<!--     } else { -->
<!--       target.style.display = 'block'; -->
<!--       this.value = 'Toogle figure'; -->
<!--     } -->
<!--   } ); -->
<!-- } ); -->
<!-- </script> -->